core/store/src/trie/mem/arena/mod.rs

```diff
@@ -0,0 +1,259 @@
+    }
+
+    /// Provides direct access to the memory.
+    pub fn raw_slice(&self) -> &'a [u8] {
+        &self.arena.raw_slice(self.pos, self.len)
+    }
+
+    /// Reads a usize at the given offset in the slice (bounds checked),
+    /// and returns it as a pointer.
+    pub fn read_ptr_at(&self, pos: usize) -> ArenaPtr<'a> {
+        let pos = usize::try_from_slice(&self.raw_slice()[pos..][..size_of::<usize>()]).unwrap();
+        ArenaPtr { arena: self.arena, pos }
+    }
+
+    /// Returns a slice within this slice. This checks bounds.
+    pub fn subslice(&self, start: usize, len: usize) -> ArenaSlice<'a> {
+        assert!(start + len <= self.len);
+        ArenaSlice { arena: self.arena, pos: self.pos + start, len }
+    }
+}
+
+/// Like `ArenaPtr` but allows writing to the memory.
+pub struct ArenaPtrMut<'a> {
+    arena: &'a mut ArenaMemory,
+    pos: usize,
+}
+
+impl<'a> ArenaPtrMut<'a> {
+    /// Makes a const copy of the pointer, which can only be used while also
+    /// holding a reference to the original pointer.
+    pub fn ptr<'b>(&'b self) -> ArenaPtr<'b> {
+        ArenaPtr { arena: self.arena, pos: self.pos }
+    }
+
+    /// Makes a mutable copy of the pointer, which can only be used while also
+    /// holding a mutable reference to the original pointer.
+    pub fn ptr_mut<'b>(&'b mut self) -> ArenaPtrMut<'b> {
+        ArenaPtrMut { arena: self.arena, pos: self.pos }
+    }
+
+    /// Makes a slice relative to the pointer, which can only be used while
+    /// also holding a mutable reference to the original pointer.
+    pub fn slice<'b>(&'b self, offset: usize, len: usize) -> ArenaSlice<'b> {
+        ArenaSlice { arena: self.arena, pos: self.pos + offset, len }
+    }
+
+    /// Like `slice` but makes a mutable slice.
+    pub fn slice_mut<'b>(&'b mut self, offset: usize, len: usize) -> ArenaSliceMut<'b> {
+        ArenaSliceMut { arena: self.arena, pos: self.pos + offset, len }
+    }
+
+    /// Provides mutable access to the whole memory, while holding a mutable
+    /// reference to the pointer.
+    pub fn arena_mut<'b>(&'b mut self) -> &'b mut ArenaMemory {
+        &mut self.arena
+    }
+}
+
+/// Represents a mutable slice of memory in the arena.
+pub struct ArenaSliceMut<'a> {
+    arena: &'a mut ArenaMemory,
+    pos: usize,
+    len: usize,
+}
+
+impl<'a> ArenaSliceMut<'a> {
+    pub fn raw_offset(&self) -> usize {
+        self.pos
+    }
+
+    pub fn len(&self) -> usize {
+        self.len
+    }
+
+    /// Writes a usize at the given offset in the slice (bounds checked).
+    pub fn write_usize_at(&mut self, pos: usize, ptr: usize) {
+        assert!(pos + 8 <= self.len);
+        ptr.serialize(&mut &mut self.arena.raw_slice_mut(self.pos + pos, 8)).unwrap();
+    }
+
+    /// Provides mutable raw memory access. It is only possible while holding
+    /// a mutable reference to the original slice.
+    pub fn raw_slice_mut<'b>(&'b mut self) -> &'b mut [u8] {
```

Comment:
Nit: The lifetimes here and in `subslice_mut` below (assuming subslice_mut compiles without, I’m always unclear with lifetime generics and defaulting) are useless, they’re the same as defaulting to the `'_` lifetime.

Reply:
subslice_mut does not actually work if we don't explicitly specify a *different* lifetime (and instead reuse `'a`). This is because, if you have a `ArenaSliceMut<'a>`, you cannot get another `ArenaSliceMut<'a>` from it. Otherwise, those are two simultaneous mutable references. What you can do however, is to have a shorter lifetime `'b` that is within `'a`, and then get an `ArenaSliceMut<'b>` from it, which is OK because you can only use the ArenaSliceMut<'b> while you are holding a `&'a mut`, meaning that the original `ArenaSliceMut<'a>` is not usable for another purpose while you hold 'b.

Reply:
Right, I didn’t mean reusing `'a`, but just not specifying the lifetime. It’d default to the `'_` lifetime, which is a function-local lifetime, just like the `'b` lifetime you defined. It’s really just a style nit, not a semantic nit :)

(And my doubt is, whether `ArenaSliceMut` actually auto-resolves to `ArenaSliceMut<'_>` or not, I’m not sure but I think it doesn’t, so `subslice_mut` would then be written like `pub fn subslice_mut(&mut self, start: usize, len: usize) -> ArenaSliceMut<'_> { ... }`)